src/excmds.ts

```diff
@@ -2511,22 +2521,24 @@ export async function fullscreen() {
 */
 //#background
 export async function tabclose(...indexes: string[]) {
-    let done
     async function maybeWinTabToTabId(id: string) {
         if (id.includes(".")) {
             const [winid, tabindex_number] = await parseWinTabIndex(id)
-            return (await browser.tabs.query({ windowId: winid, index: tabindex_number }))[0].id
+            return (await browser.tabs.query({ windowId: winid, index: tabindex_number }))[0]
         }
-        return idFromIndex(id)
+        return tabFromIndex(id)
     }
-    if (indexes.length > 0) {
-        const ids = await Promise.all(indexes.map(index => maybeWinTabToTabId(index)))
-        done = browser.tabs.remove(ids)
-    } else {
-        // Close current tab
-        done = browser.tabs.remove(await activeTabId())
+    const tabs = await Promise.all(indexes.length > 0 ? indexes.map(maybeWinTabToTabId) : [activeTab()])
+    const tabclosepinned = (await config.getAsync("tabclosepinned")) === "true"
+    if (!tabclosepinned) {
+        // Pinned tabs should not be closed, abort if one of the tabs is pinned
+        for (const tab of tabs) {
+            if (tab.pinned) {
+                throw new Error(`Tab ${tab.windowId}:${tab.index} is pinned and tabclosepinned is false, aborting tabclose`)
```

Comment:
```suggestion
                throw new Error(`Tab ${tab.windowId}:${tab.index + 1} is pinned and tabclosepinned is false, aborting tabclose`)
```

I'm not sure how to deal with `windowId` - I think it doesn't correspond to our window indexes. We could make `tabs` into a `winindex.tabindex: tabid` object but I don't think it's worth the effort